gk/main.c

```diff
@@ -1483,232 +1776,92 @@ process_pkts_front(uint16_t port_front, uint16_t port_back,
-				if (packet.flow.proto == RTE_ETHER_TYPE_IPV4) {
-					stats->tot_pkts_num_distributed++;
-					stats->tot_pkts_size_distributed +=
-						rte_pktmbuf_pkt_len(pkt);
-
-					add_pkt_acl(acl4, pkt);
-				} else if (likely(packet.flow.proto ==
-						RTE_ETHER_TYPE_IPV6)) {
-					stats->tot_pkts_num_distributed++;
-					stats->tot_pkts_size_distributed +=
-						rte_pktmbuf_pkt_len(pkt);
-
-					add_pkt_acl(acl6, pkt);
-				} else {
-					print_flow_err_msg(&packet.flow,
-						"gk: failed to get the fib entry");
-					drop_packet_front(pkt, instance);
-				}
-				continue;
-			}
+	r = num_ip_flows % batch_size;
+	q = num_ip_flows - r;
 
-			switch (fib->action) {
-			case GK_FWD_GRANTOR:
-				/*
-				 * We heuristically drop entries to
-				 * alleviate memory pressure
-				 * when the table is full.
-				 *
-				 * The entry instructs to enforce
-				 * policies over its packets,
-			 	 * initialize an entry in the
-				 * flow table, proceed as the
-				 * brand-new entry instructs, and
-			 	 * go to the next packet.
-			 	 */
-				ret = gk_hash_add_flow_entry(
-					instance, &packet.flow,
-					gk_conf->request_timeout_cycles,
-					ip_flow_hash_val, GK_REQUEST);
-				if (ret == -ENOSPC) {
-					/*
-					 * There is no room for a new
-					 * flow entry, but give this
-					 * flow a chance sending a
-					 * request to the grantor
-					 * server.
-					 */
-					struct flow_entry temp_fe;
-					initialize_flow_entry(&temp_fe,
-						&packet.flow, fib);
-					ret = gk_process_request(
-						&temp_fe, &packet,
-						gk_conf->sol_conf, stats);
-					if (ret < 0) {
-						drop_packet_front(
-							pkt, instance);
-					}
-					continue;
-				} else if (ret < 0) {
-					drop_packet_front(pkt, instance);
-					continue;
-				}
-
-				fe = &instance->ip_flow_entry_table[ret];
-				initialize_flow_entry(fe, &packet.flow, fib);
-				break;
-
-			case GK_FWD_GATEWAY_BACK_NET: {
-			 	/*
-				 * The entry instructs to forward
-				 * its packets to the gateway in
-				 * the back network, forward accordingly.
-				 *
-				 * BP block bypasses from the front to the
-				 * back interface are expected to bypass
-				 * ranges of IP addresses that should not
-				 * go through Gatekeeper.
-				 *
-				 * Notice that one needs to update
-				 * the Ethernet header.
-				 */
-
-				eth_cache = fib->u.gateway.eth_cache;
-				RTE_VERIFY(eth_cache != NULL);
-
-				if (adjust_pkt_len(pkt, back, 0) == NULL ||
-						pkt_copy_cached_eth_header(pkt,
-							eth_cache,
-							back->l2_len_out)) {
-					drop_packet_front(pkt, instance);
-					continue;
-				}
-
-				if (update_ip_hop_count(front, &packet,
-						num_pkts, icmp_bufs,
-						&instance->front_icmp_rs,
-						instance,
-						drop_packet_front) < 0)
-					continue;
-
-				tx_bufs[num_tx++] = pkt;
-				continue;
-			}
+	/* Lookup the flow entries in batches. */
+	for (i = 0; i < q; i += batch_size) {
+		ret = rte_hash_lookup_bulk(instance->ip_flow_hash_table,
```

Comment:
Rewriting this `for` loop following the example below eliminates the need of repeating the body of the loop in the following `if (r != 0)` statement.
```
done_lookups = 0;
while (done_lookups < num_ip_flows) {
	uint32_t num_keys = num_ip_flows - done_lookups;
	if (num_keys > RTE_HASH_LOOKUP_BULK_MAX)
		num_keys = RTE_HASH_LOOKUP_BULK_MAX;

	/* Do the body of the loop. */

	done_lookups += num_keys;
}
```